base/broadcast.jl

```diff
@@ -263,6 +263,7 @@ end
 
 Base.IteratorSize(::Type{<:Broadcasted{<:Any,<:NTuple{N,Base.OneTo}}}) where {N} = Base.HasShape{N}()
 Base.IteratorSize(::Type{<:Broadcasted{<:AbstractArrayStyle{N}, Nothing}}) where {N} = Base.HasShape{N}()
+Base.IteratorSize(::Type{<:Broadcasted{<:ArrayStyle, Nothing, <:Any, <:Tuple{T, N}}}) where {T, N} = Base.HasShape{ndims(T)}()
```

Comment:
This patch is too limited and not correct:
1. This only works for `Broadcasted` with 2 args.
2. We also has `ArrayConflict <: AbstractArrayStyle{Any}` in Base.
2. `ndims(T)` might not be defined (e.g. `N <: Tuple`)
3. `ndims{T}` is totaly ignored.

I suggest to extend this to:
```julia
Base.IteratorSize(::Type{T}) where {T<:Broadcasted} = Base.HasShape{ndims(T)}()
Base.ndims(bc::Broadcasted) = ndims(typeof(bc))
Base.ndims(::Type{<:Broadcasted{<:Any,<:NTuple{N,Any}}}) where {N} = N
Base.ndims(BC::Type{<:Broadcasted{<:Any,Nothing}}) = _maxndims(fieldtype(BC, 2))
function Base.ndims(BC::Type{<:Broadcasted{<:AbstractArrayStyle{N},Nothing}}) where {N}
    N isa Integer && return N
    _maxndims(fieldtype(BC, 2))
end
_maxndims(T) = mapfoldl(_ndims, max, _fieldtypes(T))
_fieldtypes(T) = ntuple(Base.Fix1(fieldtype,T), Val(fieldcount(T))) # Base.fieldtypes is not stable.
_ndims(x) = ndims(x)
_ndims(::Type{<:Tuple}) = 1
```
As for test, we'd better also add something like:
```julia
@test @inferred(Base.IteratorSize(Broadcast.broadcasted((1,2,3),a,zeros(3,3,3)))) === Base.HasShape{3}()
```
to make sure the `IteratorSize` is derived at inference stage.

Reply:
Thank you for your comments on generalising this.

> We also has ArrayConflict <: AbstractArrayStyle{Any} in Base.

I don't fully understand this point.  Could you elaborate with an example of an operation where this would come into play?

Reply:
https://github.com/JuliaLang/julia/blob/9542577c8833d905eaab40af80e9ad9636998519/test/broadcast.jl#L448-L456
```julia
@test Base.broadcasted(+, AD1(rand(3)), AD2(rand(3))) isa Broadcast.Broadcasted{Broadcast.ArrayConflict}
@test Base.broadcasted(+, AD1(rand(3)), AD2(rand(3))) isa Broadcast.Broadcasted{<:Broadcast.AbstractArrayStyle{Any}}
```